javascript/lib/types.ts

```diff
@@ -0,0 +1,328 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import { BitArray } from './bitarray';
+import { TextDecoder } from 'text-encoding';
+import { org } from './Arrow_generated';
+var arrow = org.apache.arrow;
+
+interface ArrayView {
+    slice(start: number, end: number) : ArrayView
+    toString() : string
+}
+
+export abstract class Vector {
+    name: string;
+    length: number;
+    null_count: number;
+    constructor(name: string) {
+        this.name = name;
+    }
+    /* Access datum at index i */
+    abstract get(i);
+    /* Return array representing data in the range [start, end) */
+    abstract slice(start: number, end: number);
+
+    /* Use recordBatch fieldNodes and Buffers to construct this Vector */
+    public loadData(recordBatch: any, buffer: any, bufReader: any, baseOffset: any) {
+        var fieldNode = recordBatch.nodes(bufReader.node_index);
+        this.length = fieldNode.length();
+        this.null_count = fieldNode.length();
+        bufReader.node_index += 1|0;
+
+        this.loadBuffers(recordBatch, buffer, bufReader, baseOffset);
+    }
+
+    protected abstract loadBuffers(recordBatch: any, buffer: any, bufReader: any, baseOffset: any);
+
+    /* Helper function for loading a VALIDITY buffer (for Nullable types) */
+    static loadValidityBuffer(recordBatch, buffer, bufReader, baseOffset) : BitArray {
+        var buf_meta = recordBatch.buffers(bufReader.index);
+        var offset = baseOffset + buf_meta.offset().low;
+        var length = buf_meta.length().low;
+        bufReader.index += 1|0;
+        return new BitArray(buffer, offset, length*8);
+    }
+
+    /* Helper function for loading an OFFSET buffer */
+    static loadOffsetBuffer(recordBatch, buffer, bufReader, baseOffset) : Int32Array {
+        var buf_meta = recordBatch.buffers(bufReader.index);
+        var offset = baseOffset + buf_meta.offset().low;
+        var length = buf_meta.length().low/Int32Array.BYTES_PER_ELEMENT;
+        bufReader.index += 1|0;
+        return new Int32Array(buffer, offset, length);
+    }
+
+}
+
+class SimpleVector<T extends ArrayView> extends Vector {
+    protected dataView: T;
+    private TypedArray: {new(buffer: any, offset: number, length: number) : T, BYTES_PER_ELEMENT: number};
```

Comment:
Relying on TypedArray (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray) in general is going to be very limiting. I don't have a complete view of all the tools available in JS for dealing with binary data. TypedArray classes can be used internally as a convenience to access scalar values of primitive types. 

Does the Node / JS Buffer type have the same semantics as the ArrowBuf and arrow::Buffer objects in Java and C++ respectively (reference-counted, zero-copy)? That would be really convenient if so